File: content/manual/versions/v6.1/source/includes/queryable-encryption/fundamentals/manual-encryption/manual-enc.go
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/joho/godotenv"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

const (
	KeyVaultNamespace = "encryption.__keyVault"
	DbName            = "medicalRecords"
	CollName          = "patients"
	KeyAltName        = "demo-data-key"
)

var URI string = os.Getenv("MONGODB_URI")

func main() {
	err := godotenv.Load()
	provider := "aws"
	kmsProviders := map[string]map[string]interface{}{
		provider: {
			"accessKeyId":     os.Getenv("AWS_ACCESS_KEY_ID"),
			"secretAccessKey": os.Getenv("AWS_SECRET_ACCESS_KEY"),
		},
	}
	masterKey := map[string]interface{}{
		"key":    os.Getenv("AWS_KEY_ARN"),
		"region": os.Getenv("AWS_KEY_REGION"),
	}

	// start_mongoclient
	client, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(URI))
	if err != nil {
		panic(fmt.Errorf("Client connect error %v", err))
	}
	// end_mongoclient

	// start_client_enc
	coll := client.Database(DbName).Collection(CollName)
	clientEncryptionOpts := options.ClientEncryption().SetKeyVaultNamespace(KeyVaultNamespace).SetKmsProviders(kmsProviders)
	clientEnc, err := mongo.NewClientEncryption(client, clientEncryptionOpts)
	if err != nil {
		panic(fmt.Errorf("NewClientEncryption error %v", err))
	}
	defer func() {
		_ = clientEnc.Close(context.TODO())
	}()
	// end_client_enc

	dataKeyOpts := options.DataKey().
		SetMasterKey(masterKey).
		SetKeyAltNames([]string{KeyAltName})
	dataKeyID, err := clientEnc.CreateDataKey(context.TODO(), provider, dataKeyOpts)
	if err != nil {
		panic(fmt.Errorf("create data key error %v", err))
	}

	// start_enc_and_insert
	nameRawValueType, nameRawValueData, err := bson.MarshalValue("Greg")
	if err != nil {
		panic(err)
	}
	nameRawValue := bson.RawValue{Type: nameRawValueType, Value: nameRawValueData}
	nameEncryptionOpts := options.Encrypt().
		SetAlgorithm("AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic").
		SetKeyID(dataKeyID)
	nameEncryptedField, err := clientEnc.Encrypt(
		context.TODO(),
		nameRawValue,
		nameEncryptionOpts)
	if err != nil {
		panic(err)
	}
	foodsRawValueType, foodsRawValueData, err := bson.MarshalValue(bson.A{"Grapes", "Cheese"})
	if err != nil {
		panic(err)
	}
	foodsRawValue := bson.RawValue{Type: foodsRawValueType, Value: foodsRawValueData}
	encryptionOpts := options.Encrypt().
		SetAlgorithm("AEAD_AES_256_CBC_HMAC_SHA_512-Random").
		SetKeyID(dataKeyID)
	foodsEncryptedField, err := clientEnc.Encrypt(
		context.TODO(),
		foodsRawValue,
		encryptionOpts)
	if err != nil {
		panic(err)
	}

	_, err = coll.InsertOne(
		context.TODO(),
		bson.D{{"name", nameEncryptedField}, {"foods", foodsEncryptedField}, {"age", 83}})
	if err != nil {
		panic(err)
	}
	// end_enc_and_insert

	// start_find_decrypt
	nameQueryRawValueType, nameQueryRawValueData, err := bson.MarshalValue("Greg")
	if err != nil {
		panic(err)
	}
	nameQueryRawValue := bson.RawValue{Type: nameQueryRawValueType, Value: nameQueryRawValueData}
	nameQueryEncryptionOpts := options.Encrypt().
		SetAlgorithm("AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic").
		SetKeyID(dataKeyID)
	nameQueryEncryptedField, err := clientEnc.Encrypt(
		context.TODO(),
		nameQueryRawValue,
		nameQueryEncryptionOpts)
	if err != nil {
		panic(err)
	}
	var result bson.M
	err = coll.FindOne(
		context.TODO(),
		bson.D{{"name", nameQueryEncryptedField}}).Decode(&result)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			return
		}
		panic(err)
	}
	fmt.Printf("Encrypted Document: %s\n", result)
	nameDecrypted, err := clientEnc.Decrypt(
		context.TODO(),
		result["name"].(primitive.Binary))
	foodsDecrypted, err := clientEnc.Decrypt(
		context.TODO(),
		result["foods"].(primitive.Binary))
	result["foods"] = foodsDecrypted
	result["name"] = nameDecrypted
	fmt.Printf("Decrypted Document: %s\n", result)
	// end_find_decrypt
	return
}

File: content/manual/versions/v7.0/source/includes/fundamentals/manual-encryption/manual-enc.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Threading;
using MongoDB.Driver;
using MongoDB.Bson;
using MongoDB.Driver.Encryption;

namespace Insert
{

    class InsertEncryptedDocument
    {
        public static void Main()
        {
            var connectionString = Environment.GetEnvironmentVariable("MONGODB_URI");
            var keyVaultNamespace = CollectionNamespace.FromFullName("encryption.__keyVault");
            var coll = "patients";
            var db = "medicalRecords";
            var db_namespace = $"{db}.{coll}";

            // start-credentials
            var kmsProviders = new Dictionary<string, IReadOnlyDictionary<string, object>>();
            var provider = "aws";
            var awsAccessKey = Environment.GetEnvironmentVariable("AWS_ACCESS_KEY_ID");
            var awsSecretAccessKey = Environment.GetEnvironmentVariable("AWS_SECRET_ACCESS_KEY");
            var awsKmsOptions = new Dictionary<string, object>
            {
               { "accessKeyId", awsAccessKey },
               { "secretAccessKey", awsSecretAccessKey }
            };
            kmsProviders.Add("aws", awsKmsOptions);
            var awsKeyARN = Environment.GetEnvironmentVariable("AWS_KEY_ARN"); // e.g. "arn:aws:kms:us-east-2:111122223333:alias/test-key"
            var awsKeyRegion = Environment.GetEnvironmentVariable("AWS_KEY_REGION");
            var dataKeyOptions = new DataKeyOptions(
               masterKey: new BsonDocument
               {
                   { "region", awsKeyRegion },
                   { "key", awsKeyARN },
               });
            // start_mongoclient
            var client = new MongoClient(connectionString);
            // end_mongoclient
            // start_client_enc
            var collection = client.GetDatabase(db).GetCollection<BsonDocument>(coll);
            var clientEncryptionOptions = new ClientEncryptionOptions(
                keyVaultClient: client,
                keyVaultNamespace: keyVaultNamespace,
                kmsProviders: kmsProviders);
            var clientEncryption = new ClientEncryption(clientEncryptionOptions);
            // end_client_enc

            List<string> keyNames = new List<string>();
            keyNames.Add("manual-enc-test");
            var dataKeyId = clientEncryption.CreateDataKey(provider, dataKeyOptions.With(keyNames), CancellationToken.None);

            // start_enc_and_insert
            var encryptedName = clientEncryption.Encrypt(
                "Greg",
                new EncryptOptions(algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic", keyId: dataKeyId),
                CancellationToken.None);
            var encryptedFoods = clientEncryption.Encrypt(
                new BsonArray { "Cheese", "Grapes" },
                new EncryptOptions(algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Random", keyId: dataKeyId),
                CancellationToken.None);
            collection.InsertOne(new BsonDocument { { "name", encryptedName }, { "age", 83 }, { "foods", encryptedFoods } });
            // end_enc_and_insert

            // start_find_decrypt
            var nameToQuery = "Greg";
            var encryptedNameToQuery = clientEncryption.Encrypt(
                nameToQuery,
                new EncryptOptions(algorithm: "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic", keyId: dataKeyId),
                CancellationToken.None);
            var doc = collection.Find(new BsonDocument { { "name", encryptedNameToQuery } }).Single();
            Console.WriteLine($"Encrypted document: {doc}");
            doc["name"] = clientEncryption.Decrypt(doc["name"].AsBsonBinaryData, CancellationToken.None);
            doc["foods"] = clientEncryption.Decrypt(doc["foods"].AsBsonBinaryData, CancellationToken.None);
            Console.WriteLine($"Decrypted field: {doc}");
            // end_find_decrypt
        }
    }
}
